gcc: update to 14.2.1+20250405.

oreo639 opened the pull request:
<!-- Uncomment relevant sections and delete options which are not applicable -->

#### Testing the changes
- I tested the changes in this PR: **briefly**

[ci skip]

You can test the ISOs here: https://devspace.voidlinux.org/oreo639/live/gcc14/

Known packages that need to be updated/fixed for gcc 14:
- [x] python3-lxml https://github.com/void-linux/void-packages/pull/53011
- [x] xmlto https://github.com/void-linux/void-packages/pull/53013
- [x] runit https://github.com/void-linux/void-packages/pull/52408
- [x] ntp https://github.com/void-linux/void-packages/commit/30e9e29bb5a348b2c6b7a9b3e29c13d2bc28cfae
- [x] libsasl https://github.com/void-linux/void-packages/commit/6f71c8889cc8a79411ddaca582250e19718f12da
- [x] nvi https://github.com/void-linux/void-packages/commit/a51ebf219534e5420ffc83d5a89b43dc992c9b81
- [x] xbps https://github.com/void-linux/void-packages/commit/73c8cdee1d9b05980cdc0763294cce9372612ba4
- [x] db (done in this pr)
- [x] zsh https://github.com/void-linux/void-packages/commit/1f3cb3a8c07687a61c10870b4360c0ab12e17e9f
- [x] gtk+ https://github.com/void-linux/void-packages/commit/ac57a55a5fda78618aec2ec7bb3c27404834ed04
- [x] gtk+3 https://github.com/void-linux/void-packages/pull/54187
- [x] libaom https://github.com/void-linux/void-packages/commit/fcbb94217389f64fea4c536959983814ebaadae4
- [x] perl-Tk https://github.com/void-linux/void-packages/commit/67aeb9d9b2b077d9b7f25902bbf079c1569a2633
- [x] gnu-efi-libs https://github.com/void-linux/void-packages/commit/92b2642bc8d1871605f824064863149985419010
- [x] syslinux https://github.com/void-linux/void-packages/commit/49af74a1dfb32ba1de556f4491b3870179c80872
- [x] recode https://github.com/void-linux/void-packages/pull/54707
- [x] libcddb https://github.com/void-linux/void-packages/commit/eeae3026d1356e1ac5d0547e19e2742a335ae9ca
- [x] vim https://github.com/void-linux/void-packages/pull/54706
- [x] vmaf https://github.com/void-linux/void-packages/pull/54721
- [x] audit https://github.com/void-linux/void-packages/pull/54722
- [x] aalib https://github.com/void-linux/void-packages/commit/d7769fe6d757a62fbfc0998d5cbde5dec8a7db11
- [x] cppdap (outdated vendored gtest, patch merged upstream) https://github.com/void-linux/void-packages/commit/a2c9d8ac073e9f40d6b7b1bb3df108c0d46481bb
- [x] snappy (outdated vendored gtest, upstream rejected patch stating that downstreams running their tests is not supported and that downstreams should either ship their own tests/patches or just disable them) https://github.com/void-linux/void-packages/commit/4e73161431672d245a8d558f269a74f1ab363cc8
- [x] openjdk7-bootstrap
- [x] openjdk8
- [x] gimp https://github.com/void-linux/void-packages/commit/d173951120b07bc57171f6962c92c0877becd0d8
- [x] xf86-video-nouveau https://github.com/void-linux/void-packages/pull/54720
- [x] libao https://github.com/void-linux/void-packages/commit/b9c6b010a201211c0ffba8a821f0021b9bdaf376
- [x] accountsservice https://github.com/void-linux/void-packages/commit/a1ca8cef0b2e98ba22c4d632685f85c5d0cde4c1
- Done with base-syste/xfce
- [x] zeitgeist
- [x] libgee
- [x] cogl
- [x] clutter-gst3
- [x] telepathy-glib
- [x] xen https://github.com/void-linux/void-packages/pull/54751
- [x] vde2 https://github.com/void-linux/void-packages/pull/54750
- [x] caribou https://github.com/void-linux/void-packages/commit/acb3baa4b40037dbfde90924287414eda267f0d9
- [x] mesa-demos https://github.com/void-linux/void-packages/pull/55048
- [x] libfm https://github.com/void-linux/void-packages/pull/55050
- [x] pcmanfm https://github.com/void-linux/void-packages/pull/55050
- [x] lxsession https://github.com/void-linux/void-packages/pull/55050
- [x] lxterminal https://github.com/void-linux/void-packages/pull/55050
- [x] lxpanel https://github.com/void-linux/void-packages/pull/55050
- [x] gpicview https://github.com/void-linux/void-packages/pull/55050
- [x] libaccounts-glib https://github.com/void-linux/void-packages/commit/7e1d695ed1a7e74755b42a38b5a68d58cda2dbb4
- Done with gnome/enlightenment/kde/lxde/lxqt/cinnamon/mate
- Sagemath Related
- [x] gf2x
- [x] linbox
- [x] givaro

Known packages that need to be updated for gcc 14 and musl (mostly basename):
- [x] elfutils https://github.com/void-linux/void-packages/pull/54566
- [x] iproute https://github.com/void-linux/void-packages/pull/54567
- [x] attr https://github.com/void-linux/void-packages/commit/fa60f8b037a561cbb76b44f04128d10fcbb0ceb0
- [x] kmod https://github.com/void-linux/void-packages/commit/54b0270d96cee0dc4366baefb7ad6d2a6d69aa66
- [x] sane https://github.com/void-linux/void-packages/pull/54614
- [x] mdadm https://github.com/void-linux/void-packages/commit/10694f08c471b2701b3c275540b0c665f67c8b9b
- [x] pulseaudio (broken due to configure checks detecting presence of open64, this is fixed in musl 1.2.x https://github.com/bminor/musl/commit/246f1c811448f37a44b41cd8df8d0ef9736d95f4 https://github.com/bminor/musl/commit/50ab830633134dac99011219f0210ee2759ffbde) https://github.com/void-linux/void-packages/commit/b87ec8f3ca216b6d9cfb4836f236fca13f45c2ab
- [x] gpm
- [x] linux-tools https://github.com/void-linux/void-packages/commit/8c670d7a13a3d6783e23624bda033ef898cafd72
- [x] v4l-utils (ioctl declaration) https://github.com/void-linux/void-packages/pull/54708
- [x] ffmpeg6 (ioctl declaration) https://github.com/void-linux/void-packages/pull/54708
- [x] ffmpeg (ioctl declaration) https://github.com/void-linux/void-packages/pull/54708
- [x] gst-plugins-good1 (ioctl declaration) https://github.com/void-linux/void-packages/pull/54708
- [x] libcdio (fseek64, also broken on i686-glibc) https://github.com/void-linux/void-packages/pull/54724
- [x] gcr (getpass) https://github.com/void-linux/void-packages/commit/a40788b2db81ceba7e4a9f75e1c120efd7577600
- [x] libndp https://github.com/void-linux/void-packages/commit/580546374e10b6847a66bfb4d9f27579d5112902
- [x] samba https://github.com/void-linux/void-packages/commit/0900507d38bfc7963d7b6c1a7b04341ae987f2d2
- [x] wireless_tools https://github.com/void-linux/void-packages/commit/bfd0fa934e26cdd06c32bd2573bbc007d5377ac9
- [x] efl https://github.com/void-linux/void-packages/commit/9d8b785e5308e9d4dfcc41a835289b2b3d807680
- [x] bolt https://github.com/void-linux/void-packages/commit/c0075c0e0821b116cb98ad0c3f761750d930b18f 

gcc 14 introduces some breaking changes, the most notable of which is the introduction of new C99+ errors. This needs to be tested to catch breakages (note that `-fpermissive` was also added and can be added to packages as needed).

Note that this doesn't just affect applications but also configuration tests, so some test code that might have been considered conformant C99 with gcc 13 may now be be rejected as ill-formed causing misconfiguration.

On the other hand this also results in some packages, (e.g. on musl) now erroring during compile when built without including the correct basename declaration instead of segfaulting later due to implicit pointer truncation. This is good, but ofc, packages need to be checked for this.

https://gcc.gnu.org/gcc-14/changes.html
https://gcc.gnu.org/gcc-14/porting_to.html

Note that some maintainers, like Florian Weimer, Sam James, and others, have already done a lot of work to audit packages for configuration changes and breakages: https://fedoraproject.org/wiki/Changes/PortingToModernC
If you are maintaining a package, feel free to check if there are patches related to gcc14 or the new C99 errors in [Fedora](https://src.fedoraproject.org/), [gentoo](https://gitweb.gentoo.org/repo/gentoo.git), or upstream.

If you are a package maintainer, and have the ability to, please test your packages with gcc 14 to verify no configuration changes occur.

<!--
#### New package
- This new package conforms to the [package requirements](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#package-requirements): **YES**|**NO**
-->

<!-- Note: If the build is likely to take more than 2 hours, please add ci skip tag as described in
https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#continuous-integration
and test at least one native build and, if supported, at least one cross build.
Ignore this section if this PR is not skipping CI.
-->
<!--
#### Local build testing
- I built this PR locally for my native architecture, (ARCH-LIBC)
- I built this PR locally for these architectures (if supported. mark crossbuilds):
  - aarch64-musl
  - armv7l
  - armv6l-musl
-->

Calandracas606 commented:
gcc is working for me

i've added `--enable-host-pie`, and `--enable-host-bind-now` to my build, and removed `nopie`. maybe void should consider adding those configure args too.

oreo639 commented:
> i've added `--enable-host-pie`, and `--enable-host-bind-now` to my build, and removed `nopie`. maybe void should consider adding those configure args too.

Thanks for pointing those out.

Keep in mind those flags only apply to the host executables (e.g. `/usr/bin/gcc`, `cc1`, `cc1plus`, etc. as opposed to the libraries or the default flags when compiling).
Those flags are relatively new, and looking upstream there is a patch for a runtime issue caused by `--enable-host-pie` on PPC https://github.com/gcc-mirror/gcc/commit/4cf2de9b5268224816a3d53fdd2c3d799ebfd9c8 so it might make sense to give more time for those regressions to get found and ironed out before enabling.

The only distro that enables them by default atm, afaict, is RHEL (whereas it is disabled when building gcc for Fedora).
That is just my 2 cents, feel free to let me know if I am missing something.

Calandracas606 commented:
> The only distro that enables them by default atm, afaict, is RHEL (whereas it is disabled when building gcc for Fedora). 

out of curiosity, where can one find RedHat RMP specs? I've had trouble locating them in the past

oreo639 commented:
> out of curiosity, where can one find RedHat RMP specs? I've had trouble locating them in the past

https://src.fedoraproject.org/rpms/gcc/c/20746dd85fbe810c19ef7e6dea76d3a110bb4a4d

Calandracas606 commented:
I've been using this for quite a while now.

I used this patch to work around one issue I encountered, although I think its included in newer bugfix releases

```patch
From 99a1fe6c12c733fe4923a75a79d09a66ff8abcec Mon Sep 17 00:00:00 2001
From: Jonathan Wakely <jwakely@redhat.com>
Date: Wed, 12 Jun 2024 16:47:17 +0100
Subject: [PATCH] libstdc++: Fix unwanted #pragma messages from PSTL headers
 [PR113376]

When we rebased the PSTL on upstream, in r14-2109-g3162ca09dbdc2e, a
change to how _PSTL_USAGE_WARNINGS is set was missed out, but the change
to how it's tested was included. This means that the macro is always
defined, so testing it with #ifdef (instead of using #if to test its
value) doesn't work as intended.

Revert the test to use #if again, since that part of the upstream change
was unnecessary in the first place (the macro is always defined, so
there's no need to use #ifdef to avoid -Wundef warnings).

libstdc++-v3/ChangeLog:

	PR libstdc++/113376
	* include/pstl/pstl_config.h: Use #if instead of #ifdef to test
	the _PSTL_USAGE_WARNINGS macro.
---
 libstdc++-v3/include/pstl/pstl_config.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libstdc++-v3/include/pstl/pstl_config.h b/libstdc++-v3/include/pstl/pstl_config.h
index 7157a8a492ed..6fae15460eac 100644
--- a/libstdc++-v3/include/pstl/pstl_config.h
+++ b/libstdc++-v3/include/pstl/pstl_config.h
@@ -177,7 +177,7 @@
 
 #define _PSTL_PRAGMA_MESSAGE_IMPL(x) _PSTL_PRAGMA(message(_PSTL_STRING_CONCAT(_PSTL_PRAGMA_LOCATION, x)))
 
-#if defined(_PSTL_USAGE_WARNINGS)
+#if _PSTL_USAGE_WARNINGS
 #    define _PSTL_PRAGMA_MESSAGE(x) _PSTL_PRAGMA_MESSAGE_IMPL(x)
 #    define _PSTL_PRAGMA_MESSAGE_POLICIES(x) _PSTL_PRAGMA_MESSAGE_IMPL(x)
 #else
-- 
2.43.5
```

Other than that, I've also been using the aforementioned `--enable-host-pie`, `--enable-host-bind-now`, and removed `nopie`

Vinfall commented:
Off-topic: quite surprising that Void still does not have gcc-14 :(

leahneukirchen commented:
Should we go for glibc 2.41 at the same time?

MouadCharradi commented:
With `runit-2.2.0` update has been merged.
When do you think it's a good time to merge this PR?

oreo639 commented:
I got busy and never completed testing everything I wanted (just base-system on x86_64-glibc). I will start testing again after I finish the GNOME 47 stuff and update the list accordingly.

oreo639 commented:
Thanks for testing this with the sagemath packages.
I checked it as well rn with its dependencies, so they should be fixed now.
For future reference, linbox also needs patches for GCC 15, but I'll get to that when I get to that.

oreo639 commented:
Updated to support gcc snapshots as per recommendation.

simvux commented:
Thank you so much for all your work on this <3
 
Looking forward to seeing it in my `-Su`

oreo639 commented:
Included changes from: https://github.com/void-linux/void-packages/pull/54941

oreo639 commented:
This is now merged.
Stage 1: https://github.com/void-linux/void-packages/commit/2a265ac724c986e4299a751d53318eba695accff
Stage 2: https://github.com/void-linux/void-packages/commit/cd0bcde0dce32d85760d5062937d7a71814b32f8
Stage 3: https://github.com/void-linux/void-packages/commit/ddb5871e06a1211dc71f6207d4fbaa2e12d912c9

Notes for future reference:

One issue that occured during the build was binutils requiring a version of glibc newer than what had been built yet. The builders follow a somewhat inconsistent build order (based on dependencies of one of the packages being built) and glibc is an implicit makedep (partially for bootstrapping reasons and partially due to how the cross compilers are set up). In the future, this can be avoided by either making sure to bump the shlibs minver *after* ensuring glibc has been built, or by removing the implicit makedep on glibc from binutils.
The latter option is preferable, but requires significant work.
Thankfully, @the-maldridge was able to quickly re-index the old version of binutils, and binutils could be rebuilt with the older glibc minver.

Also, a reminder that the build log will say if it is staged or not (even if grafana is unavailable), since I forgot.
